vision_doc.md

```diff
@@ -0,0 +1,279 @@
+
+```rust
+
+#[derive(Debug, Builder, Serialize)]
+#[builder(name = "ConnectionInfo", 
+    setter(into), 
+    try_setter, 
+    build_fn(skip), 
+    derive(Deserialize), 
+    preserve_attrs(serde))]
+pub struct ConnectionPool {
+    host: String,
+    
+    api_key: ApiKey,
+    
+    #[builder(default="ConnectionPool::default_timeout()")]
+    timeout: Duration,
+    
+    #[builder(setter(skip))]
+    #[serde(skip_serializing)]
+    socket: Socket
+}
+
+impl ConnectionPool {
+    /// Could be private if the builder was meant to be only way to create instance.
+    pub fn new(host: String, api_key: ApiKey, timeout: Duration) -> Result<Self> {
+        // @FAERN; validation would occur here
+        
+        let socket = unimplemented!(); // try to open the socket...
+        ConnectionPool {
+            host: host,
+            api_key: api_key,
+            timeout: timeout
+        }
+    }
+    
+    fn default_timeout() -> Duration {
+        Duration::from_millis(5000)
+    }
+}
+
+impl ConnectionInfo {
+    /// Establish a connection using the given information.
+    pub fn connect(&self) -> Result<ConnectionPool> {
+        ConnectionPool::new(
+            self.get_host()?,
+            self.get_api_key()?,
+            self.get_timeout()?)
+        )
+    }
+}
+```
+
+# Example
+```rust
+mod exporting_crate {
+    use std::net::IpAddr;
+    use serde::{Serialize, Deserialize};
+
+    #[derive(Debug, Clone, Copy, Serialize, Deserialize)]
+    #[serde(rename_all = "snake_case")]
+    pub enum Cycle {
+        Auto,
+        Slow,
+        Medium,
+        Fast,
+    }
+
+    impl FromStr for Cycle {
+        type Err = String;
+        // ... impl elided
+    }
+
+    #[derive(Debug, Clone, Serialize, Deserialize)]
+    #[serde(tag = "type", rename_all = "snake_case")]
+    pub enum Key {
+        Str(String),
+        IpAddr(IpAddr),
+    }
+
+    #[derive(Debug, Clone)]
+    pub enum Filter {
+        Approx(String),
+        Exact(Key),
+    }
+
+    impl Serialize for Filter {
+        // ... impl elided
+    }
+
+    impl Deserialize for Filter {
+        // ... impl elided
+    }
+
+    /// A data request for the Northwind metrics API.
+    /// This type is focused on modeling a well-formed stat
+    /// request, it doesn't allow for partially-built states.
+    ///
+    /// This invents a few new properties for the builder macro:
+    /// 
+    /// * `build_fn(skip)`: This tells the builder not to auto-generate the 
+    ///   build method.
+    /// * `preserve_attributes`: This tells the builder to include all #[serde] 
+    ///   attributes on the `StatRequestBuilder` fields.
+    /// * `target_factory`: Name needs work, but this tells the macro to add an 
+    ///   inherent `builder` function on the target type.
+    #[derive(Debug, Clone, Builder, Serialize)]
+    #[builder(setter(into), try_setter, preserve_attrs(serde), derive = "Deserialize", build_fn(skip), target_factory)]
+    pub struct StatRequest {
```

Comment:
Actually a bit of both. That's a wall of attributes and it's hard to understand what this is doing. `preserve_attrs(serde)` and all those serde annotations also contribute significantly to that reasoning footprint.